openedx/core/djangoapps/content/learning_sequences/api/processors/visibility.py

```diff
@@ -25,14 +26,31 @@ def usage_keys_to_remove(self, full_course_outline):
         def should_remove(visibility):
             return visibility.hide_from_toc or visibility.visible_to_staff_only
 
+        can_see_outline = self._is_active_enrolled() or self._is_outline_publicly_visible(full_course_outline)
+
         sections_to_remove = {
             sec.usage_key
             for sec in full_course_outline.sections
-            if should_remove(sec.visibility)
+            if should_remove(sec.visibility) or not can_see_outline
         }
         seqs_to_remove = {
             seq.usage_key
             for seq in full_course_outline.sequences.values()
-            if should_remove(seq.visibility)
+            if should_remove(seq.visibility) or not can_see_outline
         }
         return frozenset(sections_to_remove | seqs_to_remove)
+
+    def _is_active_enrolled(self):
+        # If user is anonymous then there is no related sets e.g courseenrollment_set available
+        # for it and it is safe to say that he/she is not enrolled in any course at all.
+        if self.user.is_anonymous:
+            return False
+
+        try:
+            enrollment = self.user.courseenrollment_set.get(course=self.course_key)
+            return enrollment.is_active
+        except ObjectDoesNotExist:  # ObjectDoesNotExist is base class of all specific DoesNotExist exceptions
+            return False
```

Comment:
Please use the [CourseEnrollment.is_enrolled](https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/models.py#L1600-L1613) method to check whether a user is enrolled. It already encapsulates the check for anonymous users and active check, and it also caches that information in the request cache.